test/Elastic.Apm.Elasticsearch.Tests/VirtualElasticsearchTests.cs

```diff
@@ -0,0 +1,171 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Elastic.Apm.Api;
+using Xunit;
+using Elastic.Apm.Tests.Mocks;
+using Elasticsearch.Net;
+using Elasticsearch.Net.VirtualizedCluster;
+using FluentAssertions;
+
+namespace Elastic.Apm.Elasticsearch.Tests
+{
+	/// <summary>
+	/// Tests the Elasticsearch APM integration against a virtual cluster that can be configured to (mis)behave
+	/// </summary>
+	public class VirtualElasticsearchTests
+	{
+		[Fact]
+		public async Task FailOverResultsInSpans()
+		{
+			var payloadSender = new MockPayloadSender();
+			var cluster = VirtualClusterWith.Nodes(10)
+				.ClientCalls(c => c.FailAlways())
+				.ClientCalls(r => r.OnPort(9209).SucceedAlways())
+				.StaticConnectionPool()
+				.AllDefaults();
+			var client = cluster.Client;
+			using (var agent = new ApmAgent(new TestAgentComponents(payloadSender: payloadSender)))
+			using (agent.Subscribe(new ElasticsearchDiagnosticsSubscriber()))
+			{
+				var searchResponse = await agent.Tracer.CaptureTransaction("Call Client", ApiConstants.ActionExec,
+					async () => await client.SearchAsync<StringResponse>(PostData.Empty)
+				);
+				searchResponse.Should().NotBeNull();
+				searchResponse.Success.Should().BeTrue();
+				searchResponse.AuditTrail.Should().NotBeEmpty();
+
+				var failed = searchResponse.AuditTrail.Where(a => a.Event == AuditEvent.BadResponse);
+				failed.Should().HaveCount(9);
+
+				var spans = payloadSender.SpansOnFirstTransaction;
+				spans.Should().NotBeEmpty();
+				var pings = spans.Where(s => s.Action == "Ping");
+				pings.Should().HaveCount(10);
+				spans.Where(s => s.Action == "CallElasticsearch").Should().HaveCount(10);
+
+				spans.Should().OnlyContain(s => s.Context != null);
+				spans.Should().OnlyContain(s => s.Context.Db != null);
+				spans.Should().OnlyContain(s => s.Context.Db.Statement != null);
+			}
+		}
+
+		[Fact]
+		public async Task ExceptionDoesNotCauseLoseOfSpan()
+		{
+			var payloadSender = new MockPayloadSender();
+			var cluster = VirtualClusterWith.Nodes(2)
+				.ClientCalls(r => r.OnPort(9200).FailAlways())
+				.ClientCalls(c => c.OnPort(9201).FailAlways(new Exception("boom!")))
+				.StaticConnectionPool()
+				.AllDefaults();
+			var client = cluster.Client;
+			using (var agent = new ApmAgent(new TestAgentComponents(payloadSender: payloadSender)))
+			using (agent.Subscribe(new ElasticsearchDiagnosticsSubscriber()))
+			{
+				try
+				{
+					var searchResponse = await agent.Tracer.CaptureTransaction("Call Client", ApiConstants.ActionExec,
+						async () => await client.SearchAsync<StringResponse>(PostData.Empty)
+					);
+					searchResponse.Should().NotBeNull();
+				}
+				catch (Exception) { }
+
+				var spans = payloadSender.SpansOnFirstTransaction;
+				spans.Should().NotBeEmpty();
+				spans.Should().OnlyContain(s => s.Context.Db.Statement != null);
```

Comment:
This seems failing. 

Debugged this a little bit, here are my findings:

For the last span `RequestPipelineDiagnosticsListener.OnResult` is called with `IApiCallDetails response`  `null`, so we hit [this line](https://github.com/elastic/apm-agent-dotnet/pull/329/files#diff-f77a4d9d52dbf901cdd0736bd4c34521R35) - therefore we don't capture any db info. 

Also I wonder about this case in general, too. Like... we set the `span.name += " (Exception)"` - but is this really an exception? Isn't this more like a case where we just don't get the info we need?

Reply:
The assertion was flawed we can never capture a statement if response is null, and response is only ever null if an exception occurred in the client. I think the span indicating this is helpful.